components/bt/common/osi/thread.c

```diff
@@ -23,389 +23,289 @@
-        if (wq->queue != 0) {
-            vQueueDelete(wq->queue);
-        }
-        wq->queue = 0;
-        wq->capacity = 0;
-        osi_free(wq);
+    if (handle) {
+        xTaskNotifyGive(handle);
     }
-    return;
 }
 
-static bool osi_thead_work_queue_get(struct work_queue *wq, struct work_item *item)
+static bool osi_thead_work_queue_get(QueueHandle_t queue, work_item_t *item)
 {
-    assert (wq != NULL);
-    assert (wq->queue != 0);
+    assert (queue != NULL);
     assert (item != NULL);
 
-    if (pdTRUE == xQueueReceive(wq->queue, item, 0)) {
+    if (xQueueReceive(queue, item, 0) == pdTRUE) {
+        OSI_TRACE_VERBOSE("queue item received");
         return true;
-    } else {
-        return false;
     }
+    OSI_TRACE_VERBOSE("queue empty");
+    return false;
 }
 
-static bool osi_thead_work_queue_put(struct work_queue *wq, const struct work_item *item, uint32_t timeout)
+static bool osi_thead_work_queue_put(QueueHandle_t queue, const work_item_t *item, uint32_t timeout_ms)
 {
-    assert (wq != NULL);
-    assert (wq->queue != 0);
+    assert (queue != NULL);
     assert (item != NULL);
 
-    bool ret = true;
-    if (timeout ==  OSI_SEM_MAX_TIMEOUT) {
-        if (xQueueSend(wq->queue, item, portMAX_DELAY) != pdTRUE) {
-            ret = false;
-        }
-    } else {
-        if (xQueueSend(wq->queue, item, timeout / portTICK_PERIOD_MS) != pdTRUE) {
-            ret = false;
-        }
+    if (xQueueSend(queue, item, osi_ms_to_ticks(timeout_ms)) == pdTRUE) {
+        return true;
     }
-
-    return ret;
+    return false;
 }
 
-static size_t osi_thead_work_queue_len(struct work_queue *wq)
+static size_t osi_thead_work_queue_len(QueueHandle_t queue)
 {
-    assert (wq != NULL);
-    assert (wq->queue != 0);
-    assert (wq->capacity != 0);
-
-    size_t available_spaces = (size_t)uxQueueSpacesAvailable(wq->queue);
+    assert (queue != NULL);
 
-    if (available_spaces <= wq->capacity) {
-        return wq->capacity - available_spaces;
-    } else {
-        assert (0);
-    }
-    return 0;
+    return (size_t)uxQueueMessagesWaiting(queue);
 }
 
 static void osi_thread_run(void *arg)
 {
-    struct osi_thread_start_arg *start = (struct osi_thread_start_arg *)arg;
-    osi_thread_t *thread = start->thread;
-
-    osi_sem_give(&start->start_sem);
-
-    while (1) {
-        int idx = 0;
-
-        osi_sem_take(&thread->work_sem, OSI_SEM_MAX_TIMEOUT);
-
-        if (thread->stop) {
-            break;
-        }
-
-        struct work_item item;
-        while (!thread->stop && idx < thread->work_queue_num) {
+    osi_thread_t *thread = (osi_thread_t*)arg;
+    work_item_t item;
+    uint8_t idx;
+
+    OSI_TRACE_DEBUG("task %p started", thread->task);
+    // signal task that we are running and clear value
+    osi_thread_task_give((TaskHandle_t)atomic_exchange(&thread->signal, 0));
+
+    do {
+        // wait to be notified through thread->task
+        osi_thread_task_take(OSI_SEM_MAX_TIMEOUT);
```

Comment:
`osi_thread_task_take` immediately follows `osi_thread_task_give`, it is probable that the task calling `osi_thread_create` cannot get notified and thus getting blocked waiting for successful creation and starting of the task, in case that the task calling osi_thread create has a lower priority.

Reply:
You are correct. To my understanding of original code, parent task (one that calls `osi_thread_create`) should wait for child task (new one, running `osi_thread_run` ) to start and child task should give `start_sem`/`signal` to parent when ready.

Before calling `osi_thread_create`, parent task sets its task handle to signal.

If parent has higher priority than child, it should create child task, reach `osi_sem_take`/`osi_thread_task_take` before child task starts and block. Scheduler switches to child task, it starts running, gives start_sem/notifies parent task and blocks in `osi_sem_take`/`osi_thread_task_take`. Scheduler switches back to parent task, semaphore/notification is given and it continues.

If parent has lower priority, creates child task and scheduler switches to child task. Child task starts, gives start_sem/signal and blocks. Scheduler switches back to parent task, it calls `osi_sem_take`/`osi_thread_task_take`, doesn't block because semaphore/task is already given and continues.

Is there some other reason for not calling `osi_thread_task_take `immediately after `osi_thread_task_give`? Maybe scheduling problems, priority problems, ...? If yes, how to avoid them?

Maybe I didn't figure out original code correctly or missed something.

Reply:
Hi @tgotic I agree with your analysis and explanation on this issue, I don't think of other reasons, the original design is to make the parent task synchronized with the operations of child task.